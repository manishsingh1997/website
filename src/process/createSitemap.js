import fs from 'fs';
import path from 'path';
import isEmpty from 'lodash/isEmpty';
import reduce from 'lodash/reduce';
import sortBy from 'lodash/sortBy';

import {getHelpNodesURLs} from 'api/sitemap';
import {FencePhotoData, GatePhotoData, DrivewayPhotoData} from 'data/photo-gallery';
import {authRoutes, galleryRoutes, helpRoutes, publicRoutes} from 'public-routes';
import config from 'website/config';

const HOME_PAGE_URL = config.publicWebsite;

const sitemapXMLTemplate = `<?xml version="1.0" encoding="UTF-8"?>
<urlset
      xmlns="http://www.sitemaps.org/schemas/sitemap/0.9"
      xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
      xsi:schemaLocation="http://www.sitemaps.org/schemas/sitemap/0.9
            http://www.sitemaps.org/schemas/sitemap/0.9/sitemap.xsd">
{{urls}}
</urlset>
`;

/**
 * Get XML-tags from string URLs array.
 * @param {string[]} urls
 */
const generateSitemap = urls => {
  const urlsAsXMLTags = reduce(
    urls,
    (xml, url) => `${xml}
      <url>
        <loc>${url}</loc>
      </url>`,
    '',
  );
  return urlsAsXMLTags;
};

/**
 * Return URLs from route config.
 * @param {{component, exact, sitemap, path}} routes
 */
const getSitemapUrls = routes => {
  const paths = routes
    .filter(route => route.sitemap !== false)
    .filter(route => !route.path.includes(':'))
    .map(route => route.path);
  return paths.map(path => `${HOME_PAGE_URL}${path}`);
};

/**
 * Get the main sitemap.
 */
export const generateErgeonSitemap = async() => {
  const urls = sortBy([
    ...getSitemapUrls(authRoutes),
    ...getSitemapUrls(publicRoutes),
    config.blogHost,
  ]);
  return generateSitemap(urls);
};

/**
 * Prepare gallery URLs for adding to the Sitemap.
 * @param {FencePhotoData|GatePhotoData|DrivewayPhotoData} galleryData
 * @param {fence|gate|driveway} type
 */
const getGalleryURLs = (galleryData, type) => {
  const urls = [];
  galleryData.forEach(
    ({categorySlug, categoryGroups}) => {
      if (!isEmpty(categoryGroups)) {
        // Add sub-categories to the Sitemap.
        categoryGroups.forEach(({groupSlug}) => {
          urls.push(`${HOME_PAGE_URL}/gallery/${type}/${categorySlug}/${groupSlug}`);
        });
      } else {
        // No sub-categories, add just the category page itself.
        urls.push(`${HOME_PAGE_URL}/gallery/${type}/${categorySlug}`);
      }
    }
  );
  return urls;
};

/**
 * Get the sitemap for the /gallery section of the website.
 */
export const generateGallerySitemap = async() => {
  const urls = sortBy([
    ...getSitemapUrls(galleryRoutes),
    ...getGalleryURLs(FencePhotoData, 'fence'),
    ...getGalleryURLs(GatePhotoData, 'gate'),
    ...getGalleryURLs(DrivewayPhotoData, 'driveway'),
  ]);
  return generateSitemap(urls);
};

/**
 * Get the sitemap for the /help section of the website.
 */
export const generateHelpSitemap = async() => {
  const urls = sortBy(
    getSitemapUrls(helpRoutes)
      .concat(await getHelpNodesURLs()),
  );
  return generateSitemap(urls);
};

/**
 * Helper function to recursively created nested dirs.
 * @param {string} filePath
 */
const createDirIfNotExists = filePath => {
  const dirname = path.dirname(filePath);
  if (fs.existsSync(dirname)) {
    return true;
  }
  createDirIfNotExists(dirname);
  fs.mkdirSync(dirname);
};

/**
 * Save a sitemap into a file.
 * @param {string[]} urls
 * @param {string} filePath
 */
const writeSiteMap = (urls, filePath) => {
  const siteMapContent = sitemapXMLTemplate.replace('{{urls}}', urls);
  createDirIfNotExists(filePath);
  fs.writeFile(filePath, siteMapContent,
    (err) => {
      if (err) throw err;
      console.log(`Sitemap ${filePath} created, please verify the content of it:`);
      console.log(siteMapContent);
    });
};

if (typeof jest !== 'undefined') {
  // Just skip in tests
} else {
  // Generate all sitemaps
  (async() => {
    writeSiteMap(await generateErgeonSitemap(), './src/data/sitemap.xml');
    writeSiteMap(await generateGallerySitemap(), './src/data/gallery/sitemap.xml');
    writeSiteMap(await generateHelpSitemap(), './src/data/help/sitemap.xml');
  })();
}
