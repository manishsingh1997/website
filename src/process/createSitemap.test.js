/* eslint-disable camelcase */

import {generateErgeonSitemap, generateGallerySitemap, generateHelpSitemap} from './createSitemap';

jest.mock('bent', () => (type) => {
  switch (type) {
    case 'json':
    default:
      // Returns help node IDs mock
      return async () => [{node_key: 1}, {node_key: 2}, {node_key: 3}];
  }
});

describe('Sitemap', () => {
  it('should generate for website', async (done) => {
    const sitemap = await generateErgeonSitemap();
    expect(sitemap).toMatchSnapshot();
    done();
  });

  it('should generate for gallery', async (done) => {
    const sitemap = await generateGallerySitemap();
    expect(sitemap).toMatchSnapshot();
    done();
  });

  it('should generate for help', async (done) => {
    const sitemap = await generateHelpSitemap();
    expect(sitemap).toMatchSnapshot();
    done();
  });
});
