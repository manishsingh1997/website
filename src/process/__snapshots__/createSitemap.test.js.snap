// Jest Snapshot v1, c

exports[`Sitemap should generate for gallery 1`] = `
"
      <url>
        <loc>http://ergeon.local:6600/gallery</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/gallery/driveway/brushed/natural</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/gallery/driveway/enhanced/enhanced</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/gallery/driveway/stamped/casual</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/gallery/driveway/stamped/modern</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/gallery/driveway/stamped/traditional</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/gallery/fence/before-after</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/gallery/fence/boxed-frame</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/gallery/fence/chain-link</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/gallery/fence/horizontal/nail-up</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/gallery/fence/horizontal/picture-frame</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/gallery/fence/horizontal/with-lattice</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/gallery/fence/lumber</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/gallery/fence/nail-up/dog-ear</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/gallery/fence/nail-up/flat-top</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/gallery/fence/nail-up/short-fence</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/gallery/fence/nail-up/with-lattice</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/gallery/fence/picture-frame/short-fence</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/gallery/fence/picture-frame/with-lattice</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/gallery/fence/picture-frame/without-lattice</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/gallery/fence/staining/semi-transparent</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/gallery/fence/staining/solid</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/gallery/fence/staining/transparent</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/gallery/gate/double/box-frame</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/gallery/gate/double/chain-link</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/gallery/gate/double/horizontal</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/gallery/gate/double/picture-frame</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/gallery/gate/double/z-frame</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/gallery/gate/single/box-frame</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/gallery/gate/single/chain-link</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/gallery/gate/single/horizontal</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/gallery/gate/single/picture-frame</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/gallery/gate/single/z-frame</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/gallery/gate/sliding</loc>
      </url>"
`;

exports[`Sitemap should generate for help 1`] = `
"
      <url>
        <loc>http://ergeon.local:6600/help</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/help/1</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/help/2</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/help/3</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/help/search</loc>
      </url>"
`;

exports[`Sitemap should generate for website 1`] = `
"
      <url>
        <loc>http://ergeon.local:6600/about-ergeon</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/careers</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/contacts</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/faq</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/licenses-warranties</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/locations</loc>
      </url>
      <url>
        <loc>http://ergeon.local:6600/request-quote</loc>
      </url>
      <url>
        <loc>https://dev.ergeon.com/blog</loc>
      </url>"
`;
